Shrink upgraded index (created on 5.x) fails in 6.x

**Elasticsearch version** (`bin/elasticsearch --version`): 6.0/ Any 6.x version

**Plugins installed**: []

**JVM version** (`java -version`): 1.8 JDK

**OS version** (`uname -a` if on a Unix-like system): Any
 
**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:

[1] Create in 5.6.x a test index

```
curl -XPOST localhost:9200/test/test -d '{ "test":true }' 
```

[2] Upgrade 5.6.x to 6.x

[3] Start Elasticsearch

[4] Run the shrinking

```
PUT test/_settings
{
  "settings": {
    "index.blocks.write": true 
  }
}

POST test/_shrink/test_shrink
{
  "settings": {
    "index.number_of_replicas": 1,
    "index.number_of_shards": 1, 
    "index.codec": "best_compression" 
  }
}
```

[5] Check the cluster health

```
{
  "cluster_name": "elasticsearch",
  "status": "red",     <----------------- IT's red one
  "timed_out": false,
  "number_of_nodes": 1,
  "number_of_data_nodes": 1,
  "active_primary_shards": 10,
  "active_shards": 10,
  "relocating_shards": 0,
  "initializing_shards": 0,
  "unassigned_shards": 11,
  "delayed_unassigned_shards": 0,
  "number_of_pending_tasks": 0,
  "number_of_in_flight_fetch": 0,
  "task_max_waiting_in_queue_millis": 0,
  "active_shards_percent_as_number": 47.61904761904761
}
```

[7] Check the RED indices

```
test_shrink                     0 p UNASSIGNED                        
test_shrink                     0 r UNASSIGNED                        
```



*Logs show the following*

```
2018-01-03T10:22:23,627][WARN ][o.e.c.a.s.ShardStateAction] [-VTajZI] [test_shrink][0] received shard failed for shard id [[test_shrink][0]], allocation id [hYCRkpmGRRC7CZCQ_As67g], primary term [0], message [failed recovery], failure [RecoveryFailedException[[test_shrink][0]: Recovery failed on {-VTajZI}{-VTajZIrRT66GVxLwD4X6w}{qW38ZFvvRFCLi0miV_YWMA}{127.0.0.1}{127.0.0.1:9300}{ml.machine_memory=17179869184, ml.max_open_jobs=20, ml.enabled=true}]; nested: IndexShardRecoveryException[failed recovery]; nested: IllegalArgumentException[Cannot use addIndexes(Directory) with indexes that have been created by a different Lucene version. The current index was generated by Lucene 7 while one of the directories contains an index that was generated with Lucene 6]; ]
org.elasticsearch.indices.recovery.RecoveryFailedException: [test_shrink][0]: Recovery failed on {-VTajZI}{-VTajZIrRT66GVxLwD4X6w}{qW38ZFvvRFCLi0miV_YWMA}{127.0.0.1}{127.0.0.1:9300}{ml.machine_memory=17179869184, ml.max_open_jobs=20, ml.enabled=true}
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$9(IndexShard.java:2077) [elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.index.shard.IndexShard$$Lambda$2882/401965331.run(Unknown Source) [elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:568) [elasticsearch-6.1.1.jar:6.1.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_45]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_45]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]
Caused by: org.elasticsearch.index.shard.IndexShardRecoveryException: failed recovery
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:334) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromLocalShards(StoreRecovery.java:122) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.index.shard.IndexShard.recoverFromLocalShards(IndexShard.java:1565) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$9(IndexShard.java:2072) ~[elasticsearch-6.1.1.jar:6.1.1]
	... 5 more
Caused by: java.lang.IllegalArgumentException: Cannot use addIndexes(Directory) with indexes that have been created by a different Lucene version. The current index was generated by Lucene 7 while one of the directories contains an index that was generated with Lucene 6
	at org.apache.lucene.index.IndexWriter.addIndexes(IndexWriter.java:2830) ~[lucene-core-7.1.0.jar:7.1.0 84c90ad2c0218156c840e19a64d72b8a38550659 - ubuntu - 2017-10-13 16:12:42]
	at org.elasticsearch.index.shard.StoreRecovery.addIndices(StoreRecovery.java:161) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromLocalShards$3(StoreRecovery.java:130) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.index.shard.StoreRecovery$$Lambda$2886/1712721504.run(Unknown Source) ~[?:?]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:292) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromLocalShards(StoreRecovery.java:122) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.index.shard.IndexShard.recoverFromLocalShards(IndexShard.java:1565) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$9(IndexShard.java:2072) ~[elasticsearch-6.1.1.jar:6.1.1]
	... 5 more
[2018-01-03T10:22:23,634][INFO ][o.e.c.r.a.AllocationService] [-VTajZI] Cluster health status changed from [YELLOW] to [RED] (reason: [shards failed [[test_shrink][0]] ...]).
```


The issue seems to be:

> Caused by: java.lang.IllegalArgumentException: Cannot use addIndexes(Directory) with indexes that have been created by a different Lucene version. The current index was generated by Lucene 7 while one of the directories contains an index that was generated with Lucene 6


~~CC @s1monw here, i think that this might be caused by https://github.com/elastic/elasticsearch/pull/22469 ?~~

--- a/core/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java
+++ b/core/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java
@@ -145,14 +145,22 @@ final class StoreRecovery {
     void addIndices(final RecoveryState.Index indexRecoveryStats, final Directory target, final Sort indexSort, final Directory[] sources,
             final long maxSeqNo, final long maxUnsafeAutoIdTimestamp, IndexMetaData indexMetaData, int shardId, boolean split,
             boolean hasNested) throws IOException {
+
+        // clean target directory (if previous recovery attempt failed) and create a fresh segment file with the proper lucene version
+        Lucene.cleanLuceneIndex(target);
+        assert sources.length > 0;
+        final int luceneIndexCreatedVersionMajor = Lucene.readSegmentInfos(sources[0]).getIndexCreatedVersionMajor();
+        new SegmentInfos(luceneIndexCreatedVersionMajor).commit(target);
+
         final Directory hardLinkOrCopyTarget = new org.apache.lucene.store.HardlinkCopyDirectoryWrapper(target);
+
         IndexWriterConfig iwc = new IndexWriterConfig(null)
             .setCommitOnClose(false)
             // we don't want merges to happen here - we call maybe merge on the engine
             // later once we stared it up otherwise we would need to wait for it here
             // we also don't specify a codec here and merges should use the engines for this index
             .setMergePolicy(NoMergePolicy.INSTANCE)
-            .setOpenMode(IndexWriterConfig.OpenMode.CREATE);
+            .setOpenMode(IndexWriterConfig.OpenMode.APPEND);
         if (indexSort != null) {
             iwc.setIndexSort(indexSort);
         }

--- a/qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java
+++ b/qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java
@@ -423,6 +423,73 @@ public class FullClusterRestartIT extends ESRestTestCase {
         assertEquals(numDocs, totalHits);
     }
 
+    public void testShrinkAfterUpgrade() throws IOException {
+        String shrunkenIndex = index + "_shrunk";
+        int numDocs;
+        if (runningAgainstOldCluster) {
+            XContentBuilder mappingsAndSettings = jsonBuilder();
+            mappingsAndSettings.startObject();
+            {
+                mappingsAndSettings.startObject("mappings");
+                mappingsAndSettings.startObject("doc");
+                mappingsAndSettings.startObject("properties");
+                {
+                    mappingsAndSettings.startObject("field");
+                    mappingsAndSettings.field("type", "text");
+                    mappingsAndSettings.endObject();
+                }
+                mappingsAndSettings.endObject();
+                mappingsAndSettings.endObject();
+                mappingsAndSettings.endObject();
+            }
+            mappingsAndSettings.endObject();
+            client().performRequest("PUT", "/" + index, Collections.emptyMap(),
+                new StringEntity(mappingsAndSettings.string(), ContentType.APPLICATION_JSON));
+
+            numDocs = randomIntBetween(512, 1024);
+            indexRandomDocuments(numDocs, true, true, i -> {
+                return JsonXContent.contentBuilder().startObject()
+                    .field("field", "value")
+                    .endObject();
+            });
+        } else {
+            String updateSettingsRequestBody = "{\"settings\": {\"index.blocks.write\": true}}";
+            Response rsp = client().performRequest("PUT", "/" + index + "/_settings", Collections.emptyMap(),
+                new StringEntity(updateSettingsRequestBody, ContentType.APPLICATION_JSON));
+            assertEquals(200, rsp.getStatusLine().getStatusCode());
+
+            String shrinkIndexRequestBody = "{\"settings\": {\"index.number_of_shards\": 1}}";
+            rsp = client().performRequest("PUT", "/" + index + "/_shrink/" + shrunkenIndex, Collections.emptyMap(),
+                new StringEntity(shrinkIndexRequestBody, ContentType.APPLICATION_JSON));
+            assertEquals(200, rsp.getStatusLine().getStatusCode());
+
+            numDocs = countOfIndexedRandomDocuments();
+        }
+
+        Response rsp = client().performRequest("POST", "/_refresh");
+        assertEquals(200, rsp.getStatusLine().getStatusCode());
+
+        Map<?, ?> response = toMap(client().performRequest("GET", "/" + index + "/_search"));
+        assertNoFailures(response);
+        int totalShards = (int) XContentMapValues.extractValue("_shards.total", response);
+        assertThat(totalShards, greaterThan(1));
+        int successfulShards = (int) XContentMapValues.extractValue("_shards.successful", response);
+        assertEquals(totalShards, successfulShards);
+        int totalHits = (int) XContentMapValues.extractValue("hits.total", response);
+        assertEquals(numDocs, totalHits);
+
+        if (runningAgainstOldCluster == false) {
+            response = toMap(client().performRequest("GET", "/" + shrunkenIndex + "/_search"));
+            assertNoFailures(response);
+            totalShards = (int) XContentMapValues.extractValue("_shards.total", response);
+            assertEquals(1, totalShards);
+            successfulShards = (int) XContentMapValues.extractValue("_shards.successful", response);
+            assertEquals(1, successfulShards);
+            totalHits = (int) XContentMapValues.extractValue("hits.total", response);
+            assertEquals(numDocs, totalHits);
+        }
+    }
+
     void assertBasicSearchWorks(int count) throws IOException {
         logger.info("--> testing basic search");
         Map<String, Object> response = toMap(client().performRequest("GET", "/" + index + "/_search"));

